qiskit/transpiler/passes/basis/basis_translator.py

```diff
@@ -85,9 +85,23 @@ def run(self, dag):
         basic_instrs = ['measure', 'reset', 'barrier', 'snapshot']
 
         target_basis = set(self._target_basis).union(basic_instrs)
-        source_basis = {(node.op.name, node.op.num_qubits)
-                        for node in dag.op_nodes()}
-
+        # source_basis is a list here so that it can accurately be checked for calibrations.
+        source_basis = [(node.op.name, node.op.num_qubits)
+                        for node in dag.op_nodes()]
+
```

Comment:
Smart -- I understand now. We have duplicates (e.g. `[('h', 1), ('h', 1), ('h', 1),]`) for all the times that an operation appears in the circuit, then we look again at all the operations that appear in the circuit, and remove ITS source basis if it has a calibration. Maybe we can make that more explicit .... Update: added suggestion (tested locally)

Reply:
Fixed in https://github.com/Qiskit/qiskit-terra/pull/5080/commits/0ce6ad2f700d92fb0a6eed324fc7506e5e988b71